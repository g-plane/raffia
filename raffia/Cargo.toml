[package]
name = "raffia"
version = "0.1.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "One parser for CSS/SCSS/Sass/Less."
repository = "https://github.com/g-plane/raffia"
license = "MIT"

[dependencies]
raffia_macro = { path = "../raffia_macro", version = "0.1" }
serde = { version = "1.0", features = ["derive"], optional = true }
smallvec = { version = "1.9", features = ["const_generics", "serde"] }

[dev-dependencies]
codespan-reporting = "0.11"
criterion = "0.3"
insta = { version = "1.19", features = ["glob", "ron"] }
lightningcss = "1.0.0-alpha.35"
serde_json = "1.0"
swc_common = "0.29.0"
swc_css_ast = "0.114.0"
swc_css_parser = "0.123.0"

[features]
serialize = ["serde"]

[[bench]]
name = "parse_selector"
harness = false

[[bench]]
name = "parse_stylesheet"
harness = false
