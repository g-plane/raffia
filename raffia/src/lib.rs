//! Raffia is a parser can parse CSS, SCSS, Sass (indented syntax) and Less.
//!
//! ## Basic Usage
//!
//! This crate provides a simple API to get started.
//!
//! First, create a parser, give it the source code and specify the syntax,
//! then call the [`parse`](Parser::parse) method:
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css); // syntax can also be `Scss`, `Sass` or `Less`
//! let result = parser.parse::<Stylesheet>();
//! match result {
//!     Ok(ast) => {
//!         // parsed successfully
//!         println!("{:#?}", ast);
//!     }
//!     Err(error) => {
//!         // it failed, error message and position can be accessed via `error`
//!         println!("{:#?}", error);
//!     }
//! }
//! ```
//!
//! ## Advanced Usage
//!
//! ### More Parser Options
//!
//! If you need to control more parser options, you can use [`ParserBuilder`].
//!
//! For example, to collect comments:
//!
//! ```rust
//! use raffia::ParserBuilder;
//!
//! let mut comments = vec![];
//! let builder = ParserBuilder::new("/* comment */ a {}").comments(&mut comments);
//! let mut parser = builder.build();
//! ```
//!
//! By default, syntax is CSS when using parser builder. You can customize it:
//!
//! ```rust
//! use raffia::{ParserBuilder, Syntax};
//!
//! let builder = ParserBuilder::new("a {}").syntax(Syntax::Scss);
//! ```
//!
//! ### Parse Partial Structure
//!
//! Sometimes you don't want to parse a full stylesheet.
//! Say you only need to parse a qualified rule or even a single declaration.
//! All you need to do is to update the generics of the [`parse`](Parser::parse) method.
//!
//! ```rust
//! use raffia::{ast::QualifiedRule, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css);
//! parser.parse::<QualifiedRule>();
//! ```
//!
//! and
//!
//! ```rust
//! use raffia::{ast::Declaration, Parser, Syntax};
//!
//! let mut parser = Parser::new("color: green", Syntax::Css);
//! parser.parse::<Declaration>();
//! ```
//!
//! Not all AST nodes support the usage above;
//! technically, those nodes that implement [`Parse`] trait are supported.
//!
//! ### Retrieve Recoverable Errors
//!
//! There may be some recoverable errors which doesn't affect on producing AST.
//! To retrieve those errors, use [`recoverable_errors`](Parser::recoverable_errors).
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("@keyframes kf { invalid {} }", Syntax::Css);
//! let result = parser.parse::<Stylesheet>();
//! assert!(result.is_ok());
//! println!("{:?}", parser.recoverable_errors());
//! ```
//!
//! ## Serialization
//!
//! Produced AST can be serialized by Serde, but this feature is disabled by default.
//! You need to enable feature `serialize` manually:
//!
//! ```toml
//! raffia = { version = "*", features = ["serialize"] }
//! ```
//!
//! Then you can pass AST to Serde.
//!
//! Note that Raffia only supports serialization. Deserialization isn't supported.

pub use config::{ParserOptions, Syntax};
pub use parser::{Parse, Parser, ParserBuilder};
pub use pos::{Span, Spanned};
pub use span_ignored_eq::SpanIgnoredEq;
pub use tokenizer::token;

pub mod ast;
mod config;
pub mod error;
mod parser;
pub mod pos;
mod span_ignored_eq;
mod tokenizer;
mod util;
