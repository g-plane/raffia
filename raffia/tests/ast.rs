use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    files::SimpleFile,
    term::{self, termcolor::Buffer},
};
use insta::{assert_ron_snapshot, glob, Settings};
use raffia::{ast::Stylesheet, Parser, Syntax};
use std::fs;

#[test]
fn ast_snapshot() {
    glob!("ast/**/*.{css,scss,sass,less}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };
        let mut parser = Parser::new(&code, syntax);
        let ast = match parser.parse::<Stylesheet>() {
            Ok(ast) => {
                let recoverable_errors = parser.recoverable_errors();
                assert!(
                    recoverable_errors.is_empty(),
                    "'{}' has recoverable errors: {recoverable_errors:?}",
                    path.file_name().unwrap().to_str().unwrap(),
                );
                ast
            }
            Err(error) => {
                let file = SimpleFile::new(path.file_name().unwrap().to_str().unwrap(), &code);
                let diagnostic = Diagnostic::error()
                    .with_message(error.kind.to_string())
                    .with_labels(vec![Label::primary((), error.span.start..error.span.end)]);
                let mut buffer = Buffer::ansi();
                let config = term::Config::default();
                term::emit(&mut buffer, &config, &file, &diagnostic).unwrap();
                panic!("\n{}", String::from_utf8(buffer.into_inner()).unwrap());
            }
        };

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_ron_snapshot!(name, ast);
        });
    });
}
