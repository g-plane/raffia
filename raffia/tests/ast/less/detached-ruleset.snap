---
source: raffia/tests/ast.rs
assertion_line: 54
---
Stylesheet(
  type: "Stylesheet",
  statements: [
    LessVariableDeclaration(
      type: "LessVariableDeclaration",
      name: LessVariable(
        type: "LessVariable",
        name: Ident(
          type: "Ident",
          name: "ruleset",
          raw: "ruleset",
          span: Span(
            start: 1,
            end: 8,
          ),
        ),
        span: Span(
          start: 0,
          end: 8,
        ),
      ),
      colonSpan: Span(
        start: 8,
        end: 9,
      ),
      value: LessDetachedRuleset(
        type: "LessDetachedRuleset",
        block: SimpleBlock(
          type: "SimpleBlock",
          statements: [
            Declaration(
              type: "Declaration",
              name: Ident(
                type: "Ident",
                name: "color",
                raw: "color",
                span: Span(
                  start: 14,
                  end: 19,
                ),
              ),
              colonSpan: Span(
                start: 19,
                end: 20,
              ),
              value: [
                Ident(
                  type: "Ident",
                  name: "black",
                  raw: "black",
                  span: Span(
                    start: 21,
                    end: 26,
                  ),
                ),
              ],
              important: None,
              lessPropertyMerge: None,
              span: Span(
                start: 14,
                end: 26,
              ),
            ),
            Declaration(
              type: "Declaration",
              name: Ident(
                type: "Ident",
                name: "background",
                raw: "background",
                span: Span(
                  start: 30,
                  end: 40,
                ),
              ),
              colonSpan: Span(
                start: 40,
                end: 41,
              ),
              value: [
                Ident(
                  type: "Ident",
                  name: "white",
                  raw: "white",
                  span: Span(
                    start: 42,
                    end: 47,
                  ),
                ),
              ],
              important: None,
              lessPropertyMerge: None,
              span: Span(
                start: 30,
                end: 47,
              ),
            ),
          ],
          span: Span(
            start: 10,
            end: 50,
          ),
        ),
        span: Span(
          start: 10,
          end: 50,
        ),
      ),
      span: Span(
        start: 0,
        end: 50,
      ),
    ),
    LessMixinCall(
      type: "LessMixinCall",
      callee: LessMixinCallee(
        type: "LessMixinCallee",
        children: [
          LessMixinCalleeChild(
            type: "LessMixinCalleeChild",
            name: ClassSelector(
              type: "ClassSelector",
              name: Ident(
                type: "Ident",
                name: "wrap-mixin",
                raw: "wrap-mixin",
                span: Span(
                  start: 53,
                  end: 63,
                ),
              ),
              span: Span(
                start: 52,
                end: 63,
              ),
            ),
            combinator: None,
            span: Span(
              start: 52,
              end: 63,
            ),
          ),
        ],
        span: Span(
          start: 52,
          end: 63,
        ),
      ),
      args: Some(LessMixinArguments(
        type: "LessMixinArguments",
        args: [
          LessDetachedRuleset(
            type: "LessDetachedRuleset",
            block: SimpleBlock(
              type: "SimpleBlock",
              statements: [
                Declaration(
                  type: "Declaration",
                  name: Ident(
                    type: "Ident",
                    name: "color",
                    raw: "color",
                    span: Span(
                      start: 68,
                      end: 73,
                    ),
                  ),
                  colonSpan: Span(
                    start: 73,
                    end: 74,
                  ),
                  value: [
                    Ident(
                      type: "Ident",
                      name: "black",
                      raw: "black",
                      span: Span(
                        start: 75,
                        end: 80,
                      ),
                    ),
                  ],
                  important: None,
                  lessPropertyMerge: None,
                  span: Span(
                    start: 68,
                    end: 80,
                  ),
                ),
                Declaration(
                  type: "Declaration",
                  name: Ident(
                    type: "Ident",
                    name: "one",
                    raw: "one",
                    span: Span(
                      start: 84,
                      end: 87,
                    ),
                  ),
                  colonSpan: Span(
                    start: 87,
                    end: 88,
                  ),
                  value: [
                    LessVariable(
                      type: "LessVariable",
                      name: Ident(
                        type: "Ident",
                        name: "a",
                        raw: "a",
                        span: Span(
                          start: 90,
                          end: 91,
                        ),
                      ),
                      span: Span(
                        start: 89,
                        end: 91,
                      ),
                    ),
                  ],
                  important: None,
                  lessPropertyMerge: None,
                  span: Span(
                    start: 84,
                    end: 91,
                  ),
                ),
                Declaration(
                  type: "Declaration",
                  name: Ident(
                    type: "Ident",
                    name: "four",
                    raw: "four",
                    span: Span(
                      start: 95,
                      end: 99,
                    ),
                  ),
                  colonSpan: Span(
                    start: 99,
                    end: 100,
                  ),
                  value: [
                    LessVariable(
                      type: "LessVariable",
                      name: Ident(
                        type: "Ident",
                        name: "d",
                        raw: "d",
                        span: Span(
                          start: 102,
                          end: 103,
                        ),
                      ),
                      span: Span(
                        start: 101,
                        end: 103,
                      ),
                    ),
                  ],
                  important: None,
                  lessPropertyMerge: None,
                  span: Span(
                    start: 95,
                    end: 103,
                  ),
                ),
              ],
              span: Span(
                start: 64,
                end: 106,
              ),
            ),
            span: Span(
              start: 64,
              end: 106,
            ),
          ),
        ],
        isCommaSeparated: true,
        separatorSpans: [],
        span: Span(
          start: 63,
          end: 107,
        ),
      )),
      important: None,
      span: Span(
        start: 52,
        end: 107,
      ),
    ),
    LessMixinCall(
      type: "LessMixinCall",
      callee: LessMixinCallee(
        type: "LessMixinCallee",
        children: [
          LessMixinCalleeChild(
            type: "LessMixinCalleeChild",
            name: ClassSelector(
              type: "ClassSelector",
              name: Ident(
                type: "Ident",
                name: "wrap-mixin",
                raw: "wrap-mixin",
                span: Span(
                  start: 111,
                  end: 121,
                ),
              ),
              span: Span(
                start: 110,
                end: 121,
              ),
            ),
            combinator: None,
            span: Span(
              start: 110,
              end: 121,
            ),
          ),
        ],
        span: Span(
          start: 110,
          end: 121,
        ),
      ),
      args: Some(LessMixinArguments(
        type: "LessMixinArguments",
        args: [
          LessMixinNamedArgument(
            type: "LessMixinNamedArgument",
            name: LessVariable(
              type: "LessVariable",
              name: Ident(
                type: "Ident",
                name: "ruleset",
                raw: "ruleset",
                span: Span(
                  start: 123,
                  end: 130,
                ),
              ),
              span: Span(
                start: 122,
                end: 130,
              ),
            ),
            colonSpan: Span(
              start: 130,
              end: 131,
            ),
            value: LessDetachedRuleset(
              type: "LessDetachedRuleset",
              block: SimpleBlock(
                type: "SimpleBlock",
                statements: [
                  Declaration(
                    type: "Declaration",
                    name: Ident(
                      type: "Ident",
                      name: "color",
                      raw: "color",
                      span: Span(
                        start: 136,
                        end: 141,
                      ),
                    ),
                    colonSpan: Span(
                      start: 141,
                      end: 142,
                    ),
                    value: [
                      Ident(
                        type: "Ident",
                        name: "red",
                        raw: "red",
                        span: Span(
                          start: 143,
                          end: 146,
                        ),
                      ),
                    ],
                    important: None,
                    lessPropertyMerge: None,
                    span: Span(
                      start: 136,
                      end: 146,
                    ),
                  ),
                ],
                span: Span(
                  start: 132,
                  end: 149,
                ),
              ),
              span: Span(
                start: 132,
                end: 149,
              ),
            ),
            span: Span(
              start: 122,
              end: 149,
            ),
          ),
        ],
        isCommaSeparated: true,
        separatorSpans: [],
        span: Span(
          start: 121,
          end: 150,
        ),
      )),
      important: None,
      span: Span(
        start: 110,
        end: 150,
      ),
    ),
    LessMixinCall(
      type: "LessMixinCall",
      callee: LessMixinCallee(
        type: "LessMixinCallee",
        children: [
          LessMixinCalleeChild(
            type: "LessMixinCalleeChild",
            name: ClassSelector(
              type: "ClassSelector",
              name: Ident(
                type: "Ident",
                name: "wrap-mixin",
                raw: "wrap-mixin",
                span: Span(
                  start: 154,
                  end: 164,
                ),
              ),
              span: Span(
                start: 153,
                end: 164,
              ),
            ),
            combinator: None,
            span: Span(
              start: 153,
              end: 164,
            ),
          ),
        ],
        span: Span(
          start: 153,
          end: 164,
        ),
      ),
      args: Some(LessMixinArguments(
        type: "LessMixinArguments",
        args: [
          LessDetachedRuleset(
            type: "LessDetachedRuleset",
            block: SimpleBlock(
              type: "SimpleBlock",
              statements: [
                Declaration(
                  type: "Declaration",
                  name: Ident(
                    type: "Ident",
                    name: "test",
                    raw: "test",
                    span: Span(
                      start: 169,
                      end: 173,
                    ),
                  ),
                  colonSpan: Span(
                    start: 173,
                    end: 174,
                  ),
                  value: [
                    Ident(
                      type: "Ident",
                      name: "extra-wrap",
                      raw: "extra-wrap",
                      span: Span(
                        start: 175,
                        end: 185,
                      ),
                    ),
                  ],
                  important: None,
                  lessPropertyMerge: None,
                  span: Span(
                    start: 169,
                    end: 185,
                  ),
                ),
                LessMixinCall(
                  type: "LessMixinCall",
                  callee: LessMixinCallee(
                    type: "LessMixinCallee",
                    children: [
                      LessMixinCalleeChild(
                        type: "LessMixinCalleeChild",
                        name: ClassSelector(
                          type: "ClassSelector",
                          name: Ident(
                            type: "Ident",
                            name: "wrap-mixin-calls-wrap",
                            raw: "wrap-mixin-calls-wrap",
                            span: Span(
                              start: 190,
                              end: 211,
                            ),
                          ),
                          span: Span(
                            start: 189,
                            end: 211,
                          ),
                        ),
                        combinator: None,
                        span: Span(
                          start: 189,
                          end: 211,
                        ),
                      ),
                    ],
                    span: Span(
                      start: 189,
                      end: 211,
                    ),
                  ),
                  args: Some(LessMixinArguments(
                    type: "LessMixinArguments",
                    args: [
                      LessDetachedRuleset(
                        type: "LessDetachedRuleset",
                        block: SimpleBlock(
                          type: "SimpleBlock",
                          statements: [
                            Declaration(
                              type: "Declaration",
                              name: Ident(
                                type: "Ident",
                                name: "test",
                                raw: "test",
                                span: Span(
                                  start: 218,
                                  end: 222,
                                ),
                              ),
                              colonSpan: Span(
                                start: 222,
                                end: 223,
                              ),
                              value: [
                                Ident(
                                  type: "Ident",
                                  name: "wrapped-twice",
                                  raw: "wrapped-twice",
                                  span: Span(
                                    start: 224,
                                    end: 237,
                                  ),
                                ),
                              ],
                              important: None,
                              lessPropertyMerge: None,
                              span: Span(
                                start: 218,
                                end: 237,
                              ),
                            ),
                          ],
                          span: Span(
                            start: 212,
                            end: 242,
                          ),
                        ),
                        span: Span(
                          start: 212,
                          end: 242,
                        ),
                      ),
                    ],
                    isCommaSeparated: true,
                    separatorSpans: [],
                    span: Span(
                      start: 211,
                      end: 243,
                    ),
                  )),
                  important: None,
                  span: Span(
                    start: 189,
                    end: 243,
                  ),
                ),
              ],
              span: Span(
                start: 165,
                end: 246,
              ),
            ),
            span: Span(
              start: 165,
              end: 246,
            ),
          ),
        ],
        isCommaSeparated: true,
        separatorSpans: [],
        span: Span(
          start: 164,
          end: 247,
        ),
      )),
      important: None,
      span: Span(
        start: 153,
        end: 247,
      ),
    ),
    LessMixinCall(
      type: "LessMixinCall",
      callee: LessMixinCallee(
        type: "LessMixinCallee",
        children: [
          LessMixinCalleeChild(
            type: "LessMixinCalleeChild",
            name: ClassSelector(
              type: "ClassSelector",
              name: Ident(
                type: "Ident",
                name: "wrap-media-mixin",
                raw: "wrap-media-mixin",
                span: Span(
                  start: 251,
                  end: 267,
                ),
              ),
              span: Span(
                start: 250,
                end: 267,
              ),
            ),
            combinator: None,
            span: Span(
              start: 250,
              end: 267,
            ),
          ),
        ],
        span: Span(
          start: 250,
          end: 267,
        ),
      ),
      args: Some(LessMixinArguments(
        type: "LessMixinArguments",
        args: [
          LessDetachedRuleset(
            type: "LessDetachedRuleset",
            block: SimpleBlock(
              type: "SimpleBlock",
              statements: [
                AtRule(
                  type: "AtRule",
                  name: Ident(
                    type: "Ident",
                    name: "media",
                    raw: "media",
                    span: Span(
                      start: 273,
                      end: 278,
                    ),
                  ),
                  prelude: Some(MediaQueryList(
                    type: "MediaQueryList",
                    queries: [
                      MediaCondition(
                        type: "MediaCondition",
                        conditions: [
                          MediaInParens(
                            type: "MediaInParens",
                            kind: MediaFeatureBoolean(
                              type: "MediaFeatureBoolean",
                              name: Ident(
                                type: "Ident",
                                name: "tv",
                                raw: "tv",
                                span: Span(
                                  start: 280,
                                  end: 282,
                                ),
                              ),
                              span: Span(
                                start: 280,
                                end: 282,
                              ),
                            ),
                            span: Span(
                              start: 279,
                              end: 283,
                            ),
                          ),
                        ],
                        span: Span(
                          start: 279,
                          end: 283,
                        ),
                      ),
                    ],
                    commaSpans: [],
                    span: Span(
                      start: 279,
                      end: 283,
                    ),
                  )),
                  block: Some(SimpleBlock(
                    type: "SimpleBlock",
                    statements: [
                      QualifiedRule(
                        type: "QualifiedRule",
                        selector: SelectorList(
                          type: "SelectorList",
                          selectors: [
                            ComplexSelector(
                              type: "ComplexSelector",
                              children: [
                                CompoundSelector(
                                  type: "CompoundSelector",
                                  children: [
                                    ClassSelector(
                                      type: "ClassSelector",
                                      name: Ident(
                                        type: "Ident",
                                        name: "triple-wrapped-mq",
                                        raw: "triple-wrapped-mq",
                                        span: Span(
                                          start: 291,
                                          end: 308,
                                        ),
                                      ),
                                      span: Span(
                                        start: 290,
                                        end: 308,
                                      ),
                                    ),
                                  ],
                                  span: Span(
                                    start: 290,
                                    end: 308,
                                  ),
                                ),
                              ],
                              span: Span(
                                start: 290,
                                end: 308,
                              ),
                            ),
                          ],
                          commaSpans: [],
                          span: Span(
                            start: 290,
                            end: 308,
                          ),
                        ),
                        block: SimpleBlock(
                          type: "SimpleBlock",
                          statements: [
                            Declaration(
                              type: "Declaration",
                              name: Ident(
                                type: "Ident",
                                name: "triple",
                                raw: "triple",
                                span: Span(
                                  start: 317,
                                  end: 323,
                                ),
                              ),
                              colonSpan: Span(
                                start: 323,
                                end: 324,
                              ),
                              value: [
                                Ident(
                                  type: "Ident",
                                  name: "true",
                                  raw: "true",
                                  span: Span(
                                    start: 325,
                                    end: 329,
                                  ),
                                ),
                              ],
                              important: None,
                              lessPropertyMerge: None,
                              span: Span(
                                start: 317,
                                end: 329,
                              ),
                            ),
                          ],
                          span: Span(
                            start: 309,
                            end: 336,
                          ),
                        ),
                        span: Span(
                          start: 290,
                          end: 336,
                        ),
                      ),
                    ],
                    span: Span(
                      start: 284,
                      end: 340,
                    ),
                  )),
                  span: Span(
                    start: 272,
                    end: 340,
                  ),
                ),
              ],
              span: Span(
                start: 268,
                end: 342,
              ),
            ),
            span: Span(
              start: 268,
              end: 342,
            ),
          ),
        ],
        isCommaSeparated: true,
        separatorSpans: [],
        span: Span(
          start: 267,
          end: 343,
        ),
      )),
      important: None,
      span: Span(
        start: 250,
        end: 343,
      ),
    ),
    LessMixinCall(
      type: "LessMixinCall",
      callee: LessMixinCallee(
        type: "LessMixinCallee",
        children: [
          LessMixinCalleeChild(
            type: "LessMixinCalleeChild",
            name: ClassSelector(
              type: "ClassSelector",
              name: Ident(
                type: "Ident",
                name: "mixin-definition",
                raw: "mixin-definition",
                span: Span(
                  start: 347,
                  end: 363,
                ),
              ),
              span: Span(
                start: 346,
                end: 363,
              ),
            ),
            combinator: None,
            span: Span(
              start: 346,
              end: 363,
            ),
          ),
        ],
        span: Span(
          start: 346,
          end: 363,
        ),
      ),
      args: Some(LessMixinArguments(
        type: "LessMixinArguments",
        args: [
          LessDetachedRuleset(
            type: "LessDetachedRuleset",
            block: SimpleBlock(
              type: "SimpleBlock",
              statements: [
                Declaration(
                  type: "Declaration",
                  name: Ident(
                    type: "Ident",
                    name: "direct",
                    raw: "direct",
                    span: Span(
                      start: 365,
                      end: 371,
                    ),
                  ),
                  colonSpan: Span(
                    start: 371,
                    end: 372,
                  ),
                  value: [
                    Ident(
                      type: "Ident",
                      name: "works",
                      raw: "works",
                      span: Span(
                        start: 373,
                        end: 378,
                      ),
                    ),
                  ],
                  important: None,
                  lessPropertyMerge: None,
                  span: Span(
                    start: 365,
                    end: 378,
                  ),
                ),
              ],
              span: Span(
                start: 364,
                end: 380,
              ),
            ),
            span: Span(
              start: 364,
              end: 380,
            ),
          ),
          LessMixinNamedArgument(
            type: "LessMixinNamedArgument",
            name: LessVariable(
              type: "LessVariable",
              name: Ident(
                type: "Ident",
                name: "b",
                raw: "b",
                span: Span(
                  start: 383,
                  end: 384,
                ),
              ),
              span: Span(
                start: 382,
                end: 384,
              ),
            ),
            colonSpan: Span(
              start: 384,
              end: 385,
            ),
            value: LessDetachedRuleset(
              type: "LessDetachedRuleset",
              block: SimpleBlock(
                type: "SimpleBlock",
                statements: [
                  Declaration(
                    type: "Declaration",
                    name: Ident(
                      type: "Ident",
                      name: "named",
                      raw: "named",
                      span: Span(
                        start: 387,
                        end: 392,
                      ),
                    ),
                    colonSpan: Span(
                      start: 392,
                      end: 393,
                    ),
                    value: [
                      Ident(
                        type: "Ident",
                        name: "works",
                        raw: "works",
                        span: Span(
                          start: 394,
                          end: 399,
                        ),
                      ),
                    ],
                    important: None,
                    lessPropertyMerge: None,
                    span: Span(
                      start: 387,
                      end: 399,
                    ),
                  ),
                ],
                span: Span(
                  start: 386,
                  end: 401,
                ),
              ),
              span: Span(
                start: 386,
                end: 401,
              ),
            ),
            span: Span(
              start: 382,
              end: 401,
            ),
          ),
        ],
        isCommaSeparated: false,
        separatorSpans: [
          Span(
            start: 380,
            end: 381,
          ),
        ],
        span: Span(
          start: 363,
          end: 402,
        ),
      )),
      important: None,
      span: Span(
        start: 346,
        end: 402,
      ),
    ),
    Function(
      type: "Function",
      name: Ident(
        type: "Ident",
        name: "each",
        raw: "each",
        span: Span(
          start: 405,
          end: 409,
        ),
      ),
      args: [
        LessVariable(
          type: "LessVariable",
          name: Ident(
            type: "Ident",
            name: "selectors",
            raw: "selectors",
            span: Span(
              start: 411,
              end: 420,
            ),
          ),
          span: Span(
            start: 410,
            end: 420,
          ),
        ),
        Delimiter(
          type: "Delimiter",
          kind: Comma,
          span: Span(
            start: 420,
            end: 421,
          ),
        ),
        LessDetachedRuleset(
          type: "LessDetachedRuleset",
          block: SimpleBlock(
            type: "SimpleBlock",
            statements: [
              QualifiedRule(
                type: "QualifiedRule",
                selector: SelectorList(
                  type: "SelectorList",
                  selectors: [
                    ComplexSelector(
                      type: "ComplexSelector",
                      children: [
                        CompoundSelector(
                          type: "CompoundSelector",
                          children: [
                            ClassSelector(
                              type: "ClassSelector",
                              name: LessInterpolatedIdent(
                                type: "LessInterpolatedIdent",
                                elements: [
                                  InterpolableIdentStaticPart(
                                    type: "InterpolableIdentStaticPart",
                                    value: "sel-",
                                    raw: "sel-",
                                    span: Span(
                                      start: 427,
                                      end: 431,
                                    ),
                                  ),
                                  LessVariableInterpolation(
                                    type: "LessVariableInterpolation",
                                    name: Ident(
                                      type: "Ident",
                                      name: "value",
                                      raw: "value",
                                      span: Span(
                                        start: 433,
                                        end: 438,
                                      ),
                                    ),
                                    span: Span(
                                      start: 431,
                                      end: 439,
                                    ),
                                  ),
                                ],
                                span: Span(
                                  start: 427,
                                  end: 439,
                                ),
                              ),
                              span: Span(
                                start: 426,
                                end: 439,
                              ),
                            ),
                          ],
                          span: Span(
                            start: 426,
                            end: 439,
                          ),
                        ),
                      ],
                      span: Span(
                        start: 426,
                        end: 439,
                      ),
                    ),
                  ],
                  commaSpans: [],
                  span: Span(
                    start: 426,
                    end: 439,
                  ),
                ),
                block: SimpleBlock(
                  type: "SimpleBlock",
                  statements: [],
                  span: Span(
                    start: 440,
                    end: 442,
                  ),
                ),
                span: Span(
                  start: 426,
                  end: 442,
                ),
              ),
            ],
            span: Span(
              start: 422,
              end: 444,
            ),
          ),
          span: Span(
            start: 422,
            end: 444,
          ),
        ),
      ],
      span: Span(
        start: 405,
        end: 445,
      ),
    ),
  ],
  span: Span(
    start: 0,
    end: 447,
  ),
)
